rspace/src/main/scala/coop/rchain/rspace/RSpace.scala

```diff
@@ -153,121 +189,179 @@ class RSpace[F[_], C, P, E, A, R, K] private[rspace] (
+                                                              c -> {
+                                                                if (c == batChannel)
+                                                                  (data, -1) +: as
+                                                                else as
+                                                              }
+                                                          )
+                                                      }
+                           result <- extractFirstMatch(
+                                      channels,
+                                      matchCandidates,
+                                      channelToIndexedDataList.toMap
+                                    ) match {
+                                      case Left(e) => e.asLeft[MaybeCandidate].pure[F]
+                                      case Right(None) =>
+                                        extractProduceCandidate(remaining, batChannel, data)
+                                      case Right(produceCandidate) =>
+                                        produceCandidate.asRight[E].pure[F]
+                                    }
+                         } yield result
+                     }
+                   for {
+                     //TODO fix double join fetch
+                     groupedChannels <- store.withTxnF(store.createTxnReadF()) { txn =>
+                                         store.getJoin(txn, channel)
+                                       }
+                     _ = logger.debug(
+                       s"""|produce: searching for matching continuations
+                           |at <groupedChannels: $groupedChannels>""".stripMargin.replace('\n', ' ')
+                     )
+                     _ = eventLog.update(produceRef +: _)
+                     extracted <- extractProduceCandidate(
+                                   groupedChannels,
+                                   channel,
+                                   Datum(data, persist, produceRef)
+                                 )
+                     r <- extracted match {
+                           case Left(e) => Left(e).pure[F]
+                           case Right(
+                               Some(
+                                 ProduceCandidate(
+                                   channels,
+                                   WaitingContinuation(
+                                     patterns,
+                                     continuation,
+                                     persistK,
+                                     consumeRef
+                                   ),
+                                   continuationIndex,
+                                   dataCandidates
+                                 )
+                               )
+                               ) =>
+                             syncF
+                               .delay {
```

Comment:
Even more so than in my previous comment, this case continuation could be expressed more clearly as a for comprehension.

Reply:
addressed in the "refactor" and "enable span" for rspace commit